core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala

```diff
@@ -398,8 +410,13 @@ final class ShuffleBlockFetcherIterator(
     var pushMergedLocalBlockBytes = 0L
     val prevNumBlocksToFetch = numBlocksToFetch
 
-    val fallback = FallbackStorage.FALLBACK_BLOCK_MANAGER_ID.executorId
-    val localExecIds = Set(blockManager.blockManagerId.executorId, fallback)
+    // Fallback to original implementation, if thread pool is not enabled.
+    val localExecIds = if (FallbackStorage.getNumReadThreads(blockManager.conf) > 0) {
```

Comment:
@maheshk114 can you explain why this change is needed?

Reply:
@attilapiros Thanks for taking a look at this PR. In the current code, the shuffle data read from local disk and external storage is handled in same way. It first tries to read from local disk and if it fails then it tries to read from external storage(BlockManager.getLocalBlockData). In this PR I have modified this behavior. The shuffle data written to external storage will be treated as remote block. If the remote block id points to external storage, then the read will be done in a separate thread. This code change is to make sure that, if fallback storage read thread pool is configured, then the fallback storage (external storage) blocks will be processed as remote block or else it will be processed as local block.

Reply:
I see that but what I do not get why we have two separate behaviour at the read side.

Reply:
I found where it was first introduced:
https://github.com/apache/spark/pull/30492/files#r532833817

@dongjoon-hyun can an externally stored block ever be found locally?

If not I would prefer to see a separate case for the external blocks and count them separately too.
So we would have the following block types:
- local
- host-local
- push-merged-local
- remote
- external

@maheshk114, @dongjoon-hyun WDYT?

Reply:
Yes that would be a proper fix.

Reply:
@dongjoon-hyun Do you think it can be  modified this way ?